extension/json/json_extract.cpp

```diff
@@ -0,0 +1,142 @@
+				                                             mask.SetInvalid(idx);
+			                                             }
+			                                             return result_val;
+		                                             });
+	} else {
+		// Columnref query
+		auto &queries = args.data[1];
+		BinaryExecutor::ExecuteWithNulls<string_t, string_t, T>(
+		    inputs, queries, result, args.size(), [&](string_t input, string_t query, ValidityMask &mask, idx_t idx) {
+			    string path;
+			    idx_t len;
+			    T result_val {};
+			    if (!JSONCommon::ConvertToPath(query, path, len) ||
+			        !TemplatedExtract<T>(input, path.c_str(), len, result_val)) {
+				    mask.SetInvalid(idx);
+			    }
+			    return result_val;
+		    });
+	}
+}
+
+static void AddFunctionAliases(vector<CreateScalarFunctionInfo> &functions, vector<string> names, ScalarFunction fun) {
+	for (const auto &name : names) {
+		fun.name = name;
+		functions.push_back(CreateScalarFunctionInfo(fun));
+	}
+}
+
+vector<CreateScalarFunctionInfo> JSONFunctions::GetExtractFunctions() {
+	vector<CreateScalarFunctionInfo> functions;
+	auto bool_fun = ScalarFunction({LogicalType::VARCHAR, LogicalType::VARCHAR}, LogicalType::BOOLEAN,
+	                               TemplatedExtractFunction<bool>, false, JSONFunctionData::Bind, nullptr, nullptr);
+	auto int_fun = ScalarFunction({LogicalType::VARCHAR, LogicalType::VARCHAR}, LogicalType::INTEGER,
+	                              TemplatedExtractFunction<int32_t>, false, JSONFunctionData::Bind, nullptr, nullptr);
+	auto bigint_fun =
+	    ScalarFunction({LogicalType::VARCHAR, LogicalType::VARCHAR}, LogicalType::BIGINT,
+	                   TemplatedExtractFunction<int64_t>, false, JSONFunctionData::Bind, nullptr, nullptr);
+	auto ubigint_fun =
+	    ScalarFunction({LogicalType::VARCHAR, LogicalType::VARCHAR}, LogicalType::UBIGINT,
+	                   TemplatedExtractFunction<uint64_t>, false, JSONFunctionData::Bind, nullptr, nullptr);
+	auto double_fun = ScalarFunction({LogicalType::VARCHAR, LogicalType::VARCHAR}, LogicalType::DOUBLE,
+	                                 TemplatedExtractFunction<double>, false, JSONFunctionData::Bind, nullptr, nullptr);
+	auto string_fun =
+	    ScalarFunction({LogicalType::VARCHAR, LogicalType::VARCHAR}, LogicalType::VARCHAR,
+	                   TemplatedExtractFunction<string_t>, false, JSONFunctionData::Bind, nullptr, nullptr);
+
+	AddFunctionAliases(functions, {"json_extract_bool", "json_extract_boolean"}, bool_fun);
+	AddFunctionAliases(functions, {"json_extract_int", "json_extract_integer"}, int_fun);
+	AddFunctionAliases(functions, {"json_extract_bigint"}, bigint_fun);
+	AddFunctionAliases(functions, {"json_extract_ubigint"}, ubigint_fun);
+	AddFunctionAliases(functions, {"json_extract_double"}, double_fun);
+	AddFunctionAliases(functions, {"json_extract_string", "json_extract_varchar"}, string_fun);
```

Comment:
Perhaps for a future PR: Most systems I have looked at also provide a generic `json_extract` function that itself returns a `json` object again, so it can be used to extract any type.

For example in SQLite:

```sql
create table json(j);
insert into json values ('{"val": "quack"}'), ('{"val": []}'), ('{"val": {"value": 1}}');
select json_extract(j, '$.val') from json;
-- quack                   
-- []                      
-- {"value":1}   
```

Or in MySQL:

```sql
create table json(j json);
insert into json values ('{"val": "quack"}'), ('{"val": []}'), ('{"val": {"value": 1}}');
select json_extract(j, '$.val') from json;
-- quack                   
-- []                      
-- {"value":1}   
```

In Postgres:

```sql
create table json(j json);
insert into json values ('{"val": "quack"}'), ('{"val": []}'), ('{"val": {"value": 1}}');
select j->'val' from json;
-- quack                   
-- []                      
-- {"value":1}   
```

In Postgres, the differentiation is made where `->` returns a JSON object, and `->>` returns a VARCHAR object. Both can be used to extract any type, but `->>` returns a VARCHAR object which can then be cast to any type that is desirable (e.g. we can cast to integer).

Reply:
I've implemented a generic `json_extract` :) Going to do the `->` syntax later and add a `JSON` LogicalType